### ISSUE TITLE
How to "daisy chain" HTTP and WS

### ISSUE THREAD

**Issue by pblemel** [warmcat/libwebsockets#1473]
Hi,

This isn't an issue, per-se but I am still having trouble accessing the mailing list so I thought I go ahead and post it here.

I have coded up a websocket client that performs the following logical steps when connecting to a server.

1) Send HTTP request to remote server requesting an authentication 'ticket' (token).
2) Parse the response from the server, saving the 'ticket'.
3) Connect to websocket server, passing ticket as URL argument
4) Communicate over the websocket, 
5) On error shut down the websocket and go back to 1 (above) to get a new ticket and try to reconnect.

I have this working intermittently using two protocols, called 'login' (HTTP) and 'client' (WS). The problem is that is extremely unreliable.  I think that the way that I have implemented it somehow interferes with the event loop.  I suspect this because of the way that I coded the interaction between the two protocols, which may be interfering with the timing of the event loop and would like to bounce it off of you.

Without posting a lot of code, here is my basic logic :

main() creates the context. My first protocol is 'login' and the second is 'client'. main() then immediately enters the lws_service loop. At this point the 'client' callback is in the driver's seat. 

1) The 'client' callback LWS_CALLBACK_PROTOCOL_INIT schedules a LWS_CALLBACK_USER callback on its wsi.
2) The 'client' LWS_CALLBACK_USER callback sets up an lws_client_connect_info struct for the login and calls lws_client_connect_via_info(). 
3) In the 'login' LWS_CALLBACK_RECEIVE_CLIENT_HTTP_READ callback I parse the auth ticket.
4) the 'login' LWS_CALLBACK_CLOSED_CLIENT_HTTP schedules a LWS_CALLBACK_USER callback

Note: Steps 1 - 4 always succeeds.

5) The 'login' LWS_CALLBACK_USER callback sets up a lws_client_connect_info struct for the websocket client and calls lws_client_connect_via_info
6) The 'client' callback sends/receives data from the websocket server.

The websocket callback successfully works its way through the connection about 20% of the time. The rest of the time the behavior that I see varies. If I let the process run to completion uninterrupted, the login works and I retrieve the ticket but the websocket connection fails (CLIENT_CONNECTION_ERROR's 'in' argument is unfortunately NULL).  If I insert a breakpoint and inspect some variables before continuing, the connection will usually succeed. For this reason I think it has to do with timing because I am effectively introducing a delay into the process, 

Before I paste in the log output for you to wade through, let me first ask, "Is this approach correct for what I am trying to do?".  Or, "Does calling lws_client_connect_via_info() to create a new connection from inside of a callback create instability in the lws event loop?"

To restate : when the client application first starts, it needs to make an HTTP request to get a ticket from the server, then append the ticket to the WS URL and start up the websocket. If the websocket ever drops for some reason, the process repeats.  This approach is necessary because I can not add an Authentication header to the websocket request (or, at least I can't figure out how to retrieve it on my server side) so I do it in the initial HTTP request.

If my approach is flawed, what is the best way to build this kind of a loop where the 'login' protocol effectively needs to interact with the 'client' protocol?  I am open to changing the plumbing if it's causing the event loop to become unstable.

I can also post my log output if you think it's worth wading through it.

Thanks,
Peter

**Comment by lws-team** [comment]
> Does calling lws_client_connect_via_info() to create a new connection from inside of a callback create instability in the lws event loop?

No... some of the minimal examples do that just fine.  "instability" is one of those words like "freeze" which gets used as  a placeholder for "I have no idea what that actually is yet".  It usually means the fruitful direction is finding out more about how those outcomes happen.

> the 'login' LWS_CALLBACK_CLOSED_CLIENT_HTTP schedules a LWS_CALLBACK_USER callback

The wsi doing the CLOSE callback is usually destroyed as soon as you return from the CLOSE callback.  If you want to create a new client connection you don't need the USER callback, you can just do it from the other guy's CLOSE callback.

Since you're making two client connections in sequence, make sure there's nothing shared between them, like using the same object to hold state etc.  The first guy has not actually been quite closed / destroyed by the time of the first guy's CLOSE comes.

Make sure you are memset-ing stack structs to 0 before use, eg, the client info structs.

Run under valgrind and look for madness.

**Comment by pblemel** [comment]
Much appreciated.  It's good to know that I'm on the right path anyway.   'unstable' means that sometimes it will connect, while other times it won't.  For what it's worth I am connecting to a J2EE server (Payara), and have Android and HTML browser applications that use the websocket reliably.

The only 'state' information shared between the two protocols is in a global character array where I store the ticket.

Here's an example of logging when the connection fails (the client is running in Fedora 17 linux).

```
[2019/01/11 01:43:16:9669] NOTICE: callback_edgeware_client: LWS_CALLBACK_USER
[2019/01/11 01:43:16:9669] USER: logging in to 172.23.93.30:80//edgeware/services/auth/login
[2019/01/11 01:43:16:9669] DEBUG: _realloc: size 472: client wsi
[2019/01/11 01:43:16:9669] INFO: lws_vhost_bind_wsi: vh default: count_bound_wsi 1
[2019/01/11 01:43:16:9670] DEBUG: lws_role_transition: 0x1b189f0: wsistate 0x10000200, ops h1
[2019/01/11 01:43:16:9670] INFO: lws_client_connect_via_info: protocol binding to edgeware-login
[2019/01/11 01:43:16:9670] DEBUG: _realloc: size 8: user space
[2019/01/11 01:43:16:9670] DEBUG: _realloc: size 64: client stash
[2019/01/11 01:43:16:9670] DEBUG: _realloc: size 13: strdup
[2019/01/11 01:43:16:9671] DEBUG: _realloc: size 30: strdup
[2019/01/11 01:43:16:9671] DEBUG: _realloc: size 16: strdup
[2019/01/11 01:43:16:9672] DEBUG: _realloc: size 16: strdup
[2019/01/11 01:43:16:9672] DEBUG: _realloc: size 15: strdup
[2019/01/11 01:43:16:9672] DEBUG: _realloc: size 5: strdup
[2019/01/11 01:43:16:9672] INFO: lws_header_table_attach: wsi 0x1b189f0: ah (nil) (tsi 0, count = 0) in
[2019/01/11 01:43:16:9673] DEBUG: _realloc: size 960: ah struct
[2019/01/11 01:43:16:9673] DEBUG: _realloc: size 4096: ah data
[2019/01/11 01:43:16:9674] INFO: _lws_create_ah: created ah 0x1b18c90 (size 4096): pool length 1
[2019/01/11 01:43:16:9674] INFO: lws_header_table_attach: did attach wsi 0x1b189f0: ah 0x1b18c90: count 1 (on exit)
[2019/01/11 01:43:16:9674] DEBUG: __lws_set_timeout: 0x1b189f0: 10 secs (reason 25)
[2019/01/11 01:43:16:9674] CLIENT: lws_client_connect_2: 0x1b189f0
[2019/01/11 01:43:16:9674] DEBUG: _realloc: size 13: strdup
[2019/01/11 01:43:16:9674] INFO: lws_client_connect_2: 0x1b189f0: address 172.23.93.30
[2019/01/11 01:43:16:9675] DEBUG: lwsi_set_state(0x1b189f0, 0x10000201)
[2019/01/11 01:43:16:9675] DEBUG: __insert_wsi_socket_into_fds: 0x1b189f0: tsi=0, sock=7, pos-in-fds=1
[2019/01/11 01:43:16:9675] DEBUG: _lws_change_pollfd: wsi 0x1b189f0: fd 7 events 1 -> 1
[2019/01/11 01:43:16:9675] DEBUG: __lws_set_timeout: 0x1b189f0: 20 secs (reason 2)
[2019/01/11 01:43:16:9676] CLIENT: nonblocking connect retry (errno = 115)
[2019/01/11 01:43:16:9676] DEBUG: _lws_change_pollfd: wsi 0x1b189f0: fd 7 events 1 -> 5
[2019/01/11 01:43:16:9678] DEBUG: _lws_change_pollfd: wsi 0x1b189f0: fd 7 events 5 -> 1
[2019/01/11 01:43:16:9678] CLIENT: lws_client_connect_2: 0x1b189f0
[2019/01/11 01:43:16:9678] INFO: lws_client_connect_2: 0x1b189f0: address 172.23.93.30
[2019/01/11 01:43:16:9678] CLIENT: connected
[2019/01/11 01:43:16:9678] INFO: lws_client_connect_2: wsi 0x1b189f0: client creating own connection
[2019/01/11 01:43:16:9678] DEBUG: lwsi_set_state(0x1b189f0, 0x10000011)
[2019/01/11 01:43:16:9678] DEBUG: __lws_set_timeout: 0x1b189f0: 20 secs (reason 8)
[2019/01/11 01:43:16:9678] DEBUG: _lws_change_pollfd: wsi 0x1b189f0: fd 7 events 1 -> 1
[2019/01/11 01:43:16:9679] DEBUG: lwsi_set_state(0x1b189f0, 0x10000012)
[2019/01/11 01:43:16:9679] DEBUG: __lws_set_timeout: 0x1b189f0: 20 secs (reason 11)
[2019/01/11 01:43:16:9679] INFO: LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER: setting username/password 'Basic cGJsZW1lbDpwYWI='
[2019/01/11 01:43:16:9679] INFO: lws_client_socket_service: HANDSHAKE2: 0x1b189f0: sending headers on 0x1b189f0 (wsistate 0x10000012 0x10000012)
[2019/01/11 01:43:16:9679] DEBUG: lwsi_set_state(0x1b189f0, 0x1000020a)
[2019/01/11 01:43:16:9679] DEBUG: __lws_set_timeout: 0x1b189f0: 20 secs (reason 4)
[2019/01/11 01:43:16:9679] DEBUG: _lws_change_pollfd: wsi 0x1b189f0: fd 7 events 1 -> 5
[2019/01/11 01:43:16:9680] DEBUG: lwsi_set_state(0x1b189f0, 0x1000020a)
[2019/01/11 01:43:16:9680] DEBUG: __lws_set_timeout: 0x1b189f0: 20 secs (reason 4)
[2019/01/11 01:43:16:9680] DEBUG: _lws_change_pollfd: wsi 0x1b189f0: fd 7 events 5 -> 1
[2019/01/11 01:43:16:9909] DEBUG: lws_client_interpret_server_handshake: 0x1b189f0: transitioning to h1 client
[2019/01/11 01:43:16:9909] DEBUG: lws_role_transition: 0x1b189f0: wsistate 0x10000117, ops h1
[2019/01/11 01:43:16:9909] DEBUG: lws_ensure_user_space: 0x1b189f0 protocol pss 8, user_space=0x1b18bd0
[2019/01/11 01:43:16:9910] INFO: lws_client_interpret_server_handshake: incoming content length 186
[2019/01/11 01:43:16:9910] DEBUG: __lws_set_timeout: 0x1b189f0: 0 secs (reason 0)
[2019/01/11 01:43:16:9910] USER: Connected with server response: 200
[2019/01/11 01:43:16:9910] INFO: lws_client_interpret_server_handshake: client connection up
[2019/01/11 01:43:16:9910] DEBUG: _lws_change_pollfd: wsi 0x1b189f0: fd 7 events 1 -> 0
[2019/01/11 01:43:16:9910] USER: RECEIVE_CLIENT_HTTP: read 0
[2019/01/11 01:43:16:9910] DEBUG: _lws_change_pollfd: wsi 0x1b189f0: fd 7 events 0 -> 1
[2019/01/11 01:43:16:9911] USER: RECEIVE_CLIENT_HTTP_READ: read 186
[2019/01/11 01:43:16:9911] USER: RECEIVE_CLIENT_HTTP_READ: ticket '63nTiuv3djUyQ3nJjF9zC_2DRX7F2dNtA_WEKw7oXFVs4aGjZ8McTY-i50OGxwpgd-HFePd9qzyP-BTKIWsy0w=='
[2019/01/11 01:43:16:9911] INFO: lws_http_transaction_completed_client: wsi: 0x1b189f0, wsi_eff: 0x1b189f0 (edgeware-login)
[2019/01/11 01:43:16:9911] USER: LWS_CALLBACK_COMPLETED_CLIENT_HTTP
[2019/01/11 01:43:16:9911] INFO: lws_callback_http_dummy: COMPLETED_CLIENT_HTTP: 0x1b189f0 (parent (nil))
[2019/01/11 01:43:16:9911] INFO: lws_http_transaction_completed_client: nothing pipelined waiting
[2019/01/11 01:43:16:9911] DEBUG: lwsi_set_state(0x1b189f0, 0x10000010)
[2019/01/11 01:43:16:9911] DEBUG: __lws_set_timeout: 0x1b189f0: 5 secs (reason 27)
[2019/01/11 01:43:22:0007] INFO: wsi 0x1b189f0: TIMEDOUT WAITING on 27 (did hdr 0, ah 0x1b18c90, wl 0, pfd events 1) 1547196202 vs 5
[2019/01/11 01:43:22:0008] INFO: __lws_close_free_wsi: 0x1b189f0: caller: timeout
[2019/01/11 01:43:22:0008] DEBUG: __lws_close_free_wsi: 0x1b189f0: DROP_PROTOCOL edgeware-login
[2019/01/11 01:43:22:0008] DEBUG: __lws_close_free_wsi: real just_kill_connection: 0x1b189f0 (sockfd 7)
[2019/01/11 01:43:22:0008] DEBUG: __remove_wsi_socket_from_fds: wsi=0x1b189f0, sock=7, fds pos=1, end guy pos=2, endfd=0
[2019/01/11 01:43:22:0008] DEBUG: lwsi_set_state(0x1b189f0, 0x1000001e)
[2019/01/11 01:43:22:0008] USER: LWS_CALLBACK_CLOSED_CLIENT_HTTP
[2019/01/11 01:43:22:0009] DEBUG: _realloc: size 40: timed_vh
[2019/01/11 01:43:22:0009] DEBUG: __lws_close_free_wsi_final: wsi 0x1b189f0: fd 7
[2019/01/11 01:43:22:0009] INFO: rops_destroy_role_h1: ah det due to close
[2019/01/11 01:43:22:0009] INFO: __lws_header_table_detach: wsi 0x1b189f0: ah 0x1b18c90 (tsi=0, count = 1)
[2019/01/11 01:43:22:0009] DEBUG: __lws_header_table_detach: wsi 0x1b189f0: ah held 6s, role/state 0x10000000 0x1e,
[2019/01/11 01:43:22:0010] INFO: __lws_header_table_detach: nobody usable waiting
[2019/01/11 01:43:22:0010] INFO: _lws_destroy_ah: freed ah 0x1b18c90 : pool length 0
[2019/01/11 01:43:22:0010] INFO: __lws_header_table_detach: wsi 0x1b189f0: ah 0x1b18c90 (tsi=0, count = 0)
[2019/01/11 01:43:22:0010] INFO: lws_vhost_unbind_wsi: vh default: count_bound_wsi 0
[2019/01/11 01:43:22:0010] DEBUG: __lws_free_wsi: 0x1b189f0, remaining wsi 1
[2019/01/11 01:43:23:0020] DEBUG: _realloc: size 472: cbwsi
[2019/01/11 01:43:23:0021] DEBUG: _realloc: size 40: cbtmr
[2019/01/11 01:43:23:0021] DEBUG: lws_service_periodic_checks: timed cb: vh default, protocol edgeware-login, reason 1000
[2019/01/11 01:43:23:0021] NOTICE: callback_edgeware_login: LWS_CALLBACK_USER
[2019/01/11 01:43:23:0022] USER: connect_edgeware_client: connecting to 172.23.93.30:80//edgeware/EPIC/data?ticket=63nTiuv3djUyQ3nJjF9zC_2DRX7F2dNtA_WEKw7oXFVs4aGjZ8McTY-i50OGxwpgd-HFePd9qzyP-BTKIWsy0w==
[2019/01/11 01:43:23:0022] DEBUG: _realloc: size 472: client wsi
[2019/01/11 01:43:23:0022] INFO: lws_vhost_bind_wsi: vh default: count_bound_wsi 1
[2019/01/11 01:43:23:0023] DEBUG: _realloc: size 200: client ws struct
[2019/01/11 01:43:23:0023] DEBUG: lws_role_transition: 0x1a4b150: wsistate 0x10000200, ops h1
[2019/01/11 01:43:23:0023] INFO: lws_client_connect_via_info: protocol binding to edgeware-client
[2019/01/11 01:43:23:0023] DEBUG: _realloc: size 4120: user space
[2019/01/11 01:43:23:0023] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_CLOSED
[2019/01/11 01:43:23:0023] INFO: lws_cancel_service
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 64: client stash
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 13: strdup
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 116: strdup
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 13: strdup
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 13: strdup
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 16: strdup
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 9: strdup
[2019/01/11 01:43:23:0024] INFO: lws_header_table_attach: wsi 0x1a4b150: ah (nil) (tsi 0, count = 0) in
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 960: ah struct
[2019/01/11 01:43:23:0024] DEBUG: _realloc: size 4096: ah data
[2019/01/11 01:43:23:0025] INFO: _lws_create_ah: created ah 0x1b18c90 (size 4096): pool length 1
[2019/01/11 01:43:23:0025] INFO: lws_header_table_attach: did attach wsi 0x1a4b150: ah 0x1b18c90: count 1 (on exit)
[2019/01/11 01:43:23:0025] DEBUG: __lws_set_timeout: 0x1a4b150: 10 secs (reason 25)
[2019/01/11 01:43:23:0025] CLIENT: lws_client_connect_2: 0x1a4b150
[2019/01/11 01:43:23:0025] DEBUG: _realloc: size 13: strdup
[2019/01/11 01:43:23:0026] INFO: lws_client_connect_2: 0x1a4b150: address 172.23.93.30
[2019/01/11 01:43:23:0026] DEBUG: lwsi_set_state(0x1a4b150, 0x10000201)
[2019/01/11 01:43:23:0026] DEBUG: __insert_wsi_socket_into_fds: 0x1a4b150: tsi=0, sock=7, pos-in-fds=1
[2019/01/11 01:43:23:0027] DEBUG: _lws_change_pollfd: wsi 0x1a4b150: fd 7 events 1 -> 1
[2019/01/11 01:43:23:0027] DEBUG: __lws_set_timeout: 0x1a4b150: 20 secs (reason 2)
[2019/01/11 01:43:23:0027] CLIENT: nonblocking connect retry (errno = 115)
[2019/01/11 01:43:23:0028] DEBUG: _lws_change_pollfd: wsi 0x1a4b150: fd 7 events 1 -> 5
[2019/01/11 01:43:23:0028] INFO: lws_context_destroy: ctx 0x1a2f160
[2019/01/11 01:43:23:0028] INFO: lws_destroy_event_pipe
[2019/01/11 01:43:23:0028] DEBUG: __remove_wsi_socket_from_fds: wsi=0x1a2f470, sock=5, fds pos=0, end guy pos=2, endfd=0
[2019/01/11 01:43:23:0028] INFO: __lws_close_free_wsi: 0x1a4b150: caller: ctx destroy
[2019/01/11 01:43:23:0028] DEBUG: __lws_close_free_wsi: 0x1a4b150: DROP_PROTOCOL edgeware-client
[2019/01/11 01:43:23:0028] ERR: callback_edgeware_client: CLIENT_CONNECTION_ERROR: (null)
[2019/01/11 01:43:23:0028] DEBUG: _realloc: size 40: timed_vh
[2019/01/11 01:43:23:0028] DEBUG: __lws_close_free_wsi: real just_kill_connection: 0x1a4b150 (sockfd 7)
[2019/01/11 01:43:23:0029] DEBUG: __remove_wsi_socket_from_fds: wsi=0x1a4b150, sock=7, fds pos=0, end guy pos=1, endfd=7
[2019/01/11 01:43:23:0029] DEBUG: lwsi_set_state(0x1a4b150, 0x1000001e)
[2019/01/11 01:43:23:0029] DEBUG: __lws_close_free_wsi_final: wsi 0x1a4b150: fd 7
[2019/01/11 01:43:23:0029] INFO: rops_destroy_role_h1: ah det due to close
[2019/01/11 01:43:23:0029] INFO: __lws_header_table_detach: wsi 0x1a4b150: ah 0x1b18c90 (tsi=0, count = 1)
[2019/01/11 01:43:23:0029] INFO: __lws_header_table_detach: nobody usable waiting
[2019/01/11 01:43:23:0029] INFO: _lws_destroy_ah: freed ah 0x1b18c90 : pool length 0
[2019/01/11 01:43:23:0029] INFO: __lws_header_table_detach: wsi 0x1a4b150: ah 0x1b18c90 (tsi=0, count = 0)
[2019/01/11 01:43:23:0029] INFO: lws_vhost_unbind_wsi: vh default: count_bound_wsi 0
[2019/01/11 01:43:23:0029] DEBUG: __lws_free_wsi: 0x1a4b150, remaining wsi 0
[2019/01/11 01:43:23:0030] INFO: lws_vhost_destroy1
[2019/01/11 01:43:23:0030] INFO: lws_context_destroy2: ctx 0x1a2f160
[2019/01/11 01:43:23:0030] NOTICE: callback_rules_fifo: LWS_CALLBACK_PROTOCOL_DESTROY
[2019/01/11 01:43:23:0030] INFO: __lws_vhost_destroy2: 0x1a489c0
[2019/01/11 01:43:23:0039] INFO:   __lws_vhost_destroy2: Freeing vhost 0x1a489c0
[2019/01/11 01:43:23:0040] INFO: lws_context_destroy3: ctx 0x1a2f160 freed
[2019/01/11 01:43:23:0040] USER: Completed failed
```                                                                             `

**Comment by lws-team** [comment]
> [2019/01/11 01:43:23:0027] CLIENT: nonblocking connect retry (errno = 115)
> [2019/01/11 01:43:23:0028] DEBUG: _lws_change_pollfd: wsi 0x1a4b150: fd 7 events 1 -> 5
> [2019/01/11 01:43:23:0028] INFO: lws_context_destroy: ctx 0x1a2f160

That doesn't look right.  You should have one context in your process, which lives for as long as any network part of your process is needed.  Both clients should be created against the same context.... is that what we're doing?  The context shouldn't be destroyed until your process is exiting, normally.

**Comment by lws-team** [comment]
`interrupted` is getting set then?  because if that's the only `lws_context_destroy()` then it means you're exiting the event loop <100us after the new client realizes it needs to go back to the event loop and wait for the connect.  If `lws_service()` is returning nonzero, that's also a big clue... maybe dump which one make it exit.

**Comment by pblemel** [comment]
Here I show my ignorance :)   The websocket client example code that I copied does the following 

```
    case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
        lwsl_err("%s: CLIENT_CONNECTION_ERROR: %s\n", __func__, in ? (char *)in : "(null)");
        vhd->client_wsi = NULL;
        schedule_callback(wsi, LWS_CALLBACK_USER, 1);
        if (*vhd->options & 1) {
            if (!*vhd->interrupted)
                *vhd->interrupted = 1;
            lws_cancel_service(lws_get_context(wsi));
        }
```
I have a breakpoint set on the server for when the connection is established.  The log above has CONNECTION_CLOSED before CONNECTION_ERROR, but the breakpoint is never triggered and the server log shows no indication that the client ever tried to connect.

If on the other hand, I set a breakpoint on the client side before attempting to connect, wait a moment, and then 'continue' the server will stop at the breakpoint on its side and the log shows the connection.  This is what I mean by it appears to be timing related.  I can generate a log for this case too, if it helps.

Thanks

**Comment by pblemel** [comment]
Thanks for putting eyes on this and your comments.  I think that what I need to do in the morning is go back through the example code and compare it to what I am doing in my code. The only thing I've added is this HTTP -> WEBSOCKET chain, but I may have inadvertently changed something during my trial-and-error trying to get things working reliably. At least I know that it's a valid approach if I didn't inadvertently break something.

I appreciate your help!
Peter

**Comment by lws-team** [comment]
```
            if (!*vhd->interrupted)
                *vhd->interrupted = 1;
```

This is specific to the function of the original minimal example... it exits the event loop under certain conditions because that's what that example wants to do.

If you set ::interrupted (which is what vhd->interrupted points to via a pvo), then the event loop exits.  Just remove that part of the code.

**Comment by pblemel** [comment]
I'll do that. I have been more focused on why the connection fails in the first place.  This change will cause the client to go back and request another ticket via HTTP after this failure, and then attempt to re-connect the websocket. It will be enlightening to see if it repeatedly fails.

**Comment by lws-team** [comment]
There's no "connection fail" in the logs... lws is nonblocking... you have to do the `connect()` twice then, once to show you want to connect and at least another time when it signals that the connection actually happened (by POLLIN or POLLOUT, can't recall which).  It's doing it the first time and returning to the event loop.  But this interrupted getting set has killed the event loop and you do the context destroy flow instead.

**Comment by pblemel** [comment]
Okay, I see that. Unless I have misunderstood what you are asking me to do, the code that we're talking about removing is in the CONNECTION_ERROR section though.  Or are you suggesting that I change something somewhere else?

**Comment by lws-team** [comment]
Get rid of anything in the callback that sets `*vhd->interrupted`... Ctrl-C will still set it if you have the SIGINT handler part from the examples, that's all you need.  The 'instability' is the callback telling the event loop to exit at the moment.

**Comment by pblemel** [comment]
Okay.  Success. No code remains that sets ```*vhd->interrupted ```.   

Success - with one final question.  What was/is happening is that when the websocket connects, CLIENT_CLOSED gets fired before CLIENT_CONNECTED for some reason - and CLIENT_CLOSED in the minimal ws client also set the interrupted bit, causing the context to be deleted. Getting rid of that code enables the client to connect, and now I see a steady stream of pings and pongs!

The question is, why is CLOSED being fired before CONNECTED? Is that normal or a sign of something else being wrong in my code?

```
[2019/01/11 03:07:09:0205] INFO: Initial logging level 1439
[2019/01/11 03:07:09:0205] INFO: Libwebsockets version: 3.0.99 v3.0.0-143-g2f8efa10
[2019/01/11 03:07:09:0206] INFO: IPV6 not compiled in
[2019/01/11 03:07:09:0206] INFO:  LWS_DEF_HEADER_LEN    : 4096
[2019/01/11 03:07:09:0206] INFO:  LWS_MAX_PROTOCOLS     : 5
[2019/01/11 03:07:09:0206] INFO:  LWS_MAX_SMP           : 1
[2019/01/11 03:07:09:0206] INFO:  sizeof (*info)        : 496
[2019/01/11 03:07:09:0206] INFO:  SYSTEM_RANDOM_FILEPATH: '/dev/urandom'
[2019/01/11 03:07:09:0206] INFO:  HTTP2 support         : not configured
[2019/01/11 03:07:09:0206] DEBUG: _realloc: size 776: context
[2019/01/11 03:07:09:0206] INFO: Using event loop: poll
[2019/01/11 03:07:09:0206] INFO: Default ALPN advertisment: http/1.1
[2019/01/11 03:07:09:0206] INFO:  default timeout (secs): 20
[2019/01/11 03:07:09:0206] DEBUG: _realloc: size 32768: pt_serv_buf
[2019/01/11 03:07:09:0206] INFO:  Threads: 1 each 1024 fds
[2019/01/11 03:07:09:0207] INFO:  mem: context:         33544 B (776 ctx + (1 thr x 32768))
[2019/01/11 03:07:09:0207] INFO:  mem: http hdr rsvd:   5177344 B (1 thr x (4096 + 960) x 1024))
[2019/01/11 03:07:09:0207] DEBUG: _realloc: size 8192: fds table
[2019/01/11 03:07:09:0207] INFO:  mem: pollfd map:       8192
[2019/01/11 03:07:09:0207] DEBUG: _realloc: size 8192: lws_lookup
[2019/01/11 03:07:09:0207] INFO:  mem: platform fd map:  8192 bytes
[2019/01/11 03:07:09:0207] DEBUG: _realloc: size 472: event pipe wsi
[2019/01/11 03:07:09:0207] DEBUG: lws_role_transition: 0x2364470: wsistate 0x200, ops pipe
[2019/01/11 03:07:09:0207] DEBUG: event pipe fd 5
[2019/01/11 03:07:09:0208] DEBUG: __insert_wsi_socket_into_fds: 0x2364470: tsi=0, sock=5, pos-in-fds=0
[2019/01/11 03:07:09:0208] INFO:  Compiled with OpenSSL support
[2019/01/11 03:07:09:0208] INFO: Doing SSL library init
[2019/01/11 03:07:09:0213] DEBUG: _realloc: size 648: create vhost
[2019/01/11 03:07:09:0213] DEBUG: _realloc: size 280: vhost-specific plugin table
[2019/01/11 03:07:09:0213] DEBUG: _realloc: size 64: same vh list
[2019/01/11 03:07:09:0213] NOTICE: Creating Vhost 'default' (serving disabled), 4 protocols, IPv6 off
[2019/01/11 03:07:09:0473] NOTICE: created client ssl context for default
[2019/01/11 03:07:09:0474] INFO:  mem: per-conn:          472 bytes + protocol rx buf
[2019/01/11 03:07:09:0474] INFO:  canonical_hostname = zaphod
[2019/01/11 03:07:09:0474] INFO: lws_cancel_service
[2019/01/11 03:07:09:0474] INFO: lws_protocol_init
[2019/01/11 03:07:09:0474] DEBUG:     vhost "default", protocol "edgeware-client", option "pipe_adapter"
[2019/01/11 03:07:09:0474] DEBUG:     vhost "default", protocol "edgeware-client", option "interrupted"
[2019/01/11 03:07:09:0474] DEBUG:     vhost "default", protocol "edgeware-client", option "port"
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-client", option "options"
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-client", option "auth_url"
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-client", option "url"
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-client", option "ads"
[2019/01/11 03:07:09:0475] USER: callback_edgeware_client: LWS_CALLBACK_PROTOCOL_INIT
[2019/01/11 03:07:09:0475] DEBUG: _realloc: size 32: protocol_vh_privs
[2019/01/11 03:07:09:0475] DEBUG: _realloc: size 80: vh priv
[2019/01/11 03:07:09:0475] DEBUG: _realloc: size 40: timed_vh
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-login", option "pipe_adapter"
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-login", option "interrupted"
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-login", option "port"
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-login", option "options"
[2019/01/11 03:07:09:0475] DEBUG:     vhost "default", protocol "edgeware-login", option "auth_url"
[2019/01/11 03:07:09:0476] DEBUG:     vhost "default", protocol "edgeware-login", option "url"
[2019/01/11 03:07:09:0476] DEBUG:     vhost "default", protocol "edgeware-login", option "ads"
[2019/01/11 03:07:09:0476] USER: callback_edgeware_login: LWS_CALLBACK_PROTOCOL_INIT
[2019/01/11 03:07:09:0476] DEBUG: _realloc: size 80: vh priv
[2019/01/11 03:07:09:0476] NOTICE: callback_rules_fifo: LWS_CALLBACK_PROTOCOL_INIT
[2019/01/11 03:07:10:0490] DEBUG: _realloc: size 472: cbwsi
[2019/01/11 03:07:10:0490] DEBUG: _realloc: size 40: cbtmr
[2019/01/11 03:07:10:0490] DEBUG: lws_service_periodic_checks: timed cb: vh default, protocol edgeware-client, reason 1000
[2019/01/11 03:07:10:0491] NOTICE: callback_edgeware_client: LWS_CALLBACK_USER
[2019/01/11 03:07:10:0491] USER: logging in to 172.23.93.30:80//edgeware/services/auth/login
[2019/01/11 03:07:10:0491] DEBUG: _realloc: size 472: client wsi
[2019/01/11 03:07:10:0491] INFO: lws_vhost_bind_wsi: vh default: count_bound_wsi 1
[2019/01/11 03:07:10:0491] DEBUG: lws_role_transition: 0x244d9f0: wsistate 0x10000200, ops h1
[2019/01/11 03:07:10:0491] INFO: lws_client_connect_via_info: protocol binding to edgeware-login
[2019/01/11 03:07:10:0491] DEBUG: _realloc: size 8: user space
[2019/01/11 03:07:10:0491] DEBUG: _realloc: size 64: client stash
[2019/01/11 03:07:10:0491] DEBUG: _realloc: size 13: strdup
[2019/01/11 03:07:10:0492] DEBUG: _realloc: size 30: strdup
[2019/01/11 03:07:10:0492] DEBUG: _realloc: size 16: strdup
[2019/01/11 03:07:10:0492] DEBUG: _realloc: size 16: strdup
[2019/01/11 03:07:10:0492] DEBUG: _realloc: size 15: strdup
[2019/01/11 03:07:10:0492] DEBUG: _realloc: size 5: strdup
[2019/01/11 03:07:10:0492] INFO: lws_header_table_attach: wsi 0x244d9f0: ah (nil) (tsi 0, count = 0) in
[2019/01/11 03:07:10:0492] DEBUG: _realloc: size 960: ah struct
[2019/01/11 03:07:10:0492] DEBUG: _realloc: size 4096: ah data
[2019/01/11 03:07:10:0493] INFO: _lws_create_ah: created ah 0x244dc90 (size 4096): pool length 1
[2019/01/11 03:07:10:0493] INFO: lws_header_table_attach: did attach wsi 0x244d9f0: ah 0x244dc90: count 1 (on exit)
[2019/01/11 03:07:10:0493] DEBUG: __lws_set_timeout: 0x244d9f0: 10 secs (reason 25)
[2019/01/11 03:07:10:0493] CLIENT: lws_client_connect_2: 0x244d9f0
[2019/01/11 03:07:10:0493] DEBUG: _realloc: size 13: strdup
[2019/01/11 03:07:10:0493] INFO: lws_client_connect_2: 0x244d9f0: address 172.23.93.30
[2019/01/11 03:07:10:0493] DEBUG: lwsi_set_state(0x244d9f0, 0x10000201)
[2019/01/11 03:07:10:0493] DEBUG: __insert_wsi_socket_into_fds: 0x244d9f0: tsi=0, sock=7, pos-in-fds=1
[2019/01/11 03:07:10:0494] DEBUG: _lws_change_pollfd: wsi 0x244d9f0: fd 7 events 1 -> 1
[2019/01/11 03:07:10:0494] DEBUG: __lws_set_timeout: 0x244d9f0: 20 secs (reason 2)
[2019/01/11 03:07:10:0494] CLIENT: nonblocking connect retry (errno = 115)
[2019/01/11 03:07:10:0494] DEBUG: _lws_change_pollfd: wsi 0x244d9f0: fd 7 events 1 -> 5
[2019/01/11 03:07:10:0643] DEBUG: _lws_change_pollfd: wsi 0x244d9f0: fd 7 events 5 -> 1
[2019/01/11 03:07:10:0643] CLIENT: lws_client_connect_2: 0x244d9f0
[2019/01/11 03:07:10:0643] INFO: lws_client_connect_2: 0x244d9f0: address 172.23.93.30
[2019/01/11 03:07:10:0643] CLIENT: connected
[2019/01/11 03:07:10:0643] INFO: lws_client_connect_2: wsi 0x244d9f0: client creating own connection
[2019/01/11 03:07:10:0643] DEBUG: lwsi_set_state(0x244d9f0, 0x10000011)
[2019/01/11 03:07:10:0644] DEBUG: __lws_set_timeout: 0x244d9f0: 20 secs (reason 8)
[2019/01/11 03:07:10:0644] DEBUG: _lws_change_pollfd: wsi 0x244d9f0: fd 7 events 1 -> 1
[2019/01/11 03:07:10:0644] DEBUG: lwsi_set_state(0x244d9f0, 0x10000012)
[2019/01/11 03:07:10:0644] DEBUG: __lws_set_timeout: 0x244d9f0: 20 secs (reason 11)
[2019/01/11 03:07:10:0644] INFO: LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER: setting username/password 'Basic cGJsZW1lbDpwYWI='
[2019/01/11 03:07:10:0644] INFO: lws_client_socket_service: HANDSHAKE2: 0x244d9f0: sending headers on 0x244d9f0 (wsistate 0x10000012 0x10000012)
[2019/01/11 03:07:10:0644] DEBUG: lwsi_set_state(0x244d9f0, 0x1000020a)
[2019/01/11 03:07:10:0644] DEBUG: __lws_set_timeout: 0x244d9f0: 20 secs (reason 4)
[2019/01/11 03:07:10:0644] DEBUG: _lws_change_pollfd: wsi 0x244d9f0: fd 7 events 1 -> 5
[2019/01/11 03:07:10:0644] DEBUG: lwsi_set_state(0x244d9f0, 0x1000020a)
[2019/01/11 03:07:10:0645] DEBUG: __lws_set_timeout: 0x244d9f0: 20 secs (reason 4)
[2019/01/11 03:07:10:0645] DEBUG: _lws_change_pollfd: wsi 0x244d9f0: fd 7 events 5 -> 1
[2019/01/11 03:07:10:1535] DEBUG: lws_client_interpret_server_handshake: 0x244d9f0: transitioning to h1 client
[2019/01/11 03:07:10:1535] DEBUG: lws_role_transition: 0x244d9f0: wsistate 0x10000117, ops h1
[2019/01/11 03:07:10:1535] DEBUG: lws_ensure_user_space: 0x244d9f0 protocol pss 8, user_space=0x244dbd0
[2019/01/11 03:07:10:1535] INFO: lws_client_interpret_server_handshake: incoming content length 186
[2019/01/11 03:07:10:1535] DEBUG: __lws_set_timeout: 0x244d9f0: 0 secs (reason 0)
[2019/01/11 03:07:10:1535] USER: Connected with server response: 200
[2019/01/11 03:07:10:1535] INFO: lws_client_interpret_server_handshake: client connection up
[2019/01/11 03:07:10:1535] DEBUG: _lws_change_pollfd: wsi 0x244d9f0: fd 7 events 1 -> 0
[2019/01/11 03:07:10:1535] USER: RECEIVE_CLIENT_HTTP: read 0
[2019/01/11 03:07:10:1535] DEBUG: _lws_change_pollfd: wsi 0x244d9f0: fd 7 events 0 -> 1
[2019/01/11 03:07:10:1536] USER: RECEIVE_CLIENT_HTTP_READ: read 186
[2019/01/11 03:07:10:1536] USER: RECEIVE_CLIENT_HTTP_READ: ticket 'cuKkLU4_gyhvaRqDFCBTlrnAR13j30gSY4VZwmT9-ZyMqeZMEcqp76D0hx3zFe93jaYUTEY0VSWDreG3WDRgJQ=='
[2019/01/11 03:07:10:1536] INFO: lws_http_transaction_completed_client: wsi: 0x244d9f0, wsi_eff: 0x244d9f0 (edgeware-login)
[2019/01/11 03:07:10:1536] USER: LWS_CALLBACK_COMPLETED_CLIENT_HTTP
[2019/01/11 03:07:10:1536] INFO: lws_callback_http_dummy: COMPLETED_CLIENT_HTTP: 0x244d9f0 (parent (nil))
[2019/01/11 03:07:10:1536] INFO: lws_http_transaction_completed_client: nothing pipelined waiting
[2019/01/11 03:07:10:1536] DEBUG: lwsi_set_state(0x244d9f0, 0x10000010)
[2019/01/11 03:07:10:1536] DEBUG: __lws_set_timeout: 0x244d9f0: 5 secs (reason 27)
[2019/01/11 03:07:16:0007] INFO: wsi 0x244d9f0: TIMEDOUT WAITING on 27 (did hdr 0, ah 0x244dc90, wl 0, pfd events 1) 1547201236 vs 5
[2019/01/11 03:07:16:0008] INFO: __lws_close_free_wsi: 0x244d9f0: caller: timeout
[2019/01/11 03:07:16:0008] DEBUG: __lws_close_free_wsi: 0x244d9f0: DROP_PROTOCOL edgeware-login
[2019/01/11 03:07:16:0008] DEBUG: __lws_close_free_wsi: real just_kill_connection: 0x244d9f0 (sockfd 7)
[2019/01/11 03:07:16:0008] DEBUG: __remove_wsi_socket_from_fds: wsi=0x244d9f0, sock=7, fds pos=1, end guy pos=2, endfd=0
[2019/01/11 03:07:16:0008] DEBUG: lwsi_set_state(0x244d9f0, 0x1000001e)
[2019/01/11 03:07:16:0009] USER: LWS_CALLBACK_CLOSED_CLIENT_HTTP
[2019/01/11 03:07:16:0009] DEBUG: _realloc: size 40: timed_vh
[2019/01/11 03:07:16:0009] DEBUG: __lws_close_free_wsi_final: wsi 0x244d9f0: fd 7
[2019/01/11 03:07:16:0009] INFO: rops_destroy_role_h1: ah det due to close
[2019/01/11 03:07:16:0009] INFO: __lws_header_table_detach: wsi 0x244d9f0: ah 0x244dc90 (tsi=0, count = 1)
[2019/01/11 03:07:16:0009] DEBUG: __lws_header_table_detach: wsi 0x244d9f0: ah held 6s, role/state 0x10000000 0x1e,
[2019/01/11 03:07:16:0009] INFO: __lws_header_table_detach: nobody usable waiting
[2019/01/11 03:07:16:0009] INFO: _lws_destroy_ah: freed ah 0x244dc90 : pool length 0
[2019/01/11 03:07:16:0009] INFO: __lws_header_table_detach: wsi 0x244d9f0: ah 0x244dc90 (tsi=0, count = 0)
[2019/01/11 03:07:16:0010] INFO: lws_vhost_unbind_wsi: vh default: count_bound_wsi 0
[2019/01/11 03:07:16:0010] DEBUG: __lws_free_wsi: 0x244d9f0, remaining wsi 1
[2019/01/11 03:07:17:0047] DEBUG: _realloc: size 472: cbwsi
[2019/01/11 03:07:17:0047] DEBUG: _realloc: size 40: cbtmr
[2019/01/11 03:07:17:0047] DEBUG: lws_service_periodic_checks: timed cb: vh default, protocol edgeware-login, reason 1000
[2019/01/11 03:07:17:0047] NOTICE: callback_edgeware_login: LWS_CALLBACK_USER
[2019/01/11 03:07:17:0048] USER: connect_edgeware_client: connecting to 172.23.93.30:80//edgeware/EPIC/data?ticket=cuKkLU4_gyhvaRqDFCBTlrnAR13j30gSY4VZwmT9-ZyMqeZMEcqp76D0hx3zFe93jaYUTEY0VSWDreG3WDRgJQ==
[2019/01/11 03:07:17:0048] DEBUG: _realloc: size 472: client wsi
[2019/01/11 03:07:17:0048] INFO: lws_vhost_bind_wsi: vh default: count_bound_wsi 1
[2019/01/11 03:07:17:0048] DEBUG: _realloc: size 200: client ws struct
[2019/01/11 03:07:17:0048] DEBUG: lws_role_transition: 0x2380150: wsistate 0x10000200, ops h1
[2019/01/11 03:07:17:0048] INFO: lws_client_connect_via_info: protocol binding to edgeware-client
[2019/01/11 03:07:17:0048] DEBUG: _realloc: size 4120: user space
[2019/01/11 03:07:17:0048] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_CLOSED
[2019/01/11 03:07:17:0048] DEBUG: _realloc: size 64: client stash
[2019/01/11 03:07:17:0048] DEBUG: _realloc: size 13: strdup
[2019/01/11 03:07:17:0048] DEBUG: _realloc: size 116: strdup
[2019/01/11 03:07:17:0048] DEBUG: _realloc: size 13: strdup
[2019/01/11 03:07:17:0049] DEBUG: _realloc: size 13: strdup
[2019/01/11 03:07:17:0049] DEBUG: _realloc: size 16: strdup
[2019/01/11 03:07:17:0049] DEBUG: _realloc: size 9: strdup
[2019/01/11 03:07:17:0049] INFO: lws_header_table_attach: wsi 0x2380150: ah (nil) (tsi 0, count = 0) in
[2019/01/11 03:07:17:0049] DEBUG: _realloc: size 960: ah struct
[2019/01/11 03:07:17:0049] DEBUG: _realloc: size 4096: ah data
[2019/01/11 03:07:17:0049] INFO: _lws_create_ah: created ah 0x244dc90 (size 4096): pool length 1
[2019/01/11 03:07:17:0049] INFO: lws_header_table_attach: did attach wsi 0x2380150: ah 0x244dc90: count 1 (on exit)
[2019/01/11 03:07:17:0049] DEBUG: __lws_set_timeout: 0x2380150: 10 secs (reason 25)
[2019/01/11 03:07:17:0049] CLIENT: lws_client_connect_2: 0x2380150
[2019/01/11 03:07:17:0049] DEBUG: _realloc: size 13: strdup
[2019/01/11 03:07:17:0049] INFO: lws_client_connect_2: 0x2380150: address 172.23.93.30
[2019/01/11 03:07:17:0050] DEBUG: lwsi_set_state(0x2380150, 0x10000201)
[2019/01/11 03:07:17:0050] DEBUG: __insert_wsi_socket_into_fds: 0x2380150: tsi=0, sock=7, pos-in-fds=1
[2019/01/11 03:07:17:0050] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 1
[2019/01/11 03:07:17:0050] DEBUG: __lws_set_timeout: 0x2380150: 20 secs (reason 2)
[2019/01/11 03:07:17:0050] CLIENT: nonblocking connect retry (errno = 115)
[2019/01/11 03:07:17:0050] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 5
[2019/01/11 03:07:17:0127] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 5 -> 1
[2019/01/11 03:07:17:0127] CLIENT: lws_client_connect_2: 0x2380150
[2019/01/11 03:07:17:0128] INFO: lws_client_connect_2: 0x2380150: address 172.23.93.30
[2019/01/11 03:07:17:0128] CLIENT: connected
[2019/01/11 03:07:17:0128] INFO: lws_client_connect_2: wsi 0x2380150: client creating own connection
[2019/01/11 03:07:17:0128] DEBUG: lwsi_set_state(0x2380150, 0x10000011)
[2019/01/11 03:07:17:0128] DEBUG: __lws_set_timeout: 0x2380150: 20 secs (reason 8)
[2019/01/11 03:07:17:0128] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 1
[2019/01/11 03:07:17:0128] DEBUG: lwsi_set_state(0x2380150, 0x10000012)
[2019/01/11 03:07:17:0128] DEBUG: __lws_set_timeout: 0x2380150: 20 secs (reason 11)
[2019/01/11 03:07:17:0128] INFO: lws_client_socket_service: HANDSHAKE2: 0x2380150: sending headers on 0x2380150 (wsistate 0x10000012 0x10000012)
[2019/01/11 03:07:17:0129] DEBUG: lwsi_set_state(0x2380150, 0x1000020a)
[2019/01/11 03:07:17:0129] DEBUG: __lws_set_timeout: 0x2380150: 20 secs (reason 4)
[2019/01/11 03:07:17:0129] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 5
[2019/01/11 03:07:17:0129] DEBUG: lwsi_set_state(0x2380150, 0x1000020a)
[2019/01/11 03:07:17:0129] DEBUG: __lws_set_timeout: 0x2380150: 20 secs (reason 4)
[2019/01/11 03:07:17:0129] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 5 -> 1
[2019/01/11 03:07:17:0357] DEBUG: Selected protocol edgeware-client
[2019/01/11 03:07:17:0357] DEBUG: lws_ensure_user_space: 0x2380150 protocol pss 4120, user_space=0x244f150
[2019/01/11 03:07:17:0357] DEBUG: __lws_set_timeout: 0x2380150: 0 secs (reason 0)
[2019/01/11 03:07:17:0357] INFO: __lws_header_table_detach: wsi 0x2380150: ah 0x244dc90 (tsi=0, count = 1)
[2019/01/11 03:07:17:0357] INFO: __lws_header_table_detach: nobody usable waiting
[2019/01/11 03:07:17:0357] INFO: _lws_destroy_ah: freed ah 0x244dc90 : pool length 0
[2019/01/11 03:07:17:0358] INFO: __lws_header_table_detach: wsi 0x2380150: ah 0x244dc90 (tsi=0, count = 0)
[2019/01/11 03:07:17:0358] DEBUG: lws_role_transition: 0x2380150: wsistate 0x10000117, ops ws
[2019/01/11 03:07:17:0358] DEBUG: _realloc: size 4116: client frame buffer
[2019/01/11 03:07:17:0358] INFO: Allocating client RX buffer 4112
[2019/01/11 03:07:17:0358] DEBUG: handshake OK for protocol edgeware-client
[2019/01/11 03:07:17:0358] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_ESTABLISHED
[2019/01/11 03:07:17:0359] DEBUG: _realloc: size 32: ring create
[2019/01/11 03:07:17:0359] DEBUG: _realloc: size 24576: ring buf
[2019/01/11 03:07:22:0364] USER: callback_edgeware_client: LWS_CALLBACK_TIMER
[2019/01/11 03:07:22:0365] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 5
[2019/01/11 03:07:22:0365] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 5 -> 1
[2019/01/11 03:07:22:0365] DEBUG: lws_handle_POLLOUT_event: 0x2380150: non mux: wsistate 0x10000117, ops ws
[2019/01/11 03:07:22:0365] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_WRITEABLE
[2019/01/11 03:07:22:0365] USER: callback_edgeware_client:  (nothing in ring)
[2019/01/11 03:07:22:0365] INFO: lws_rx_flow_control: 0x2380150 0x1
[2019/01/11 03:07:22:0365] INFO: lws_rx_flow_control: 0x2380150: bitmap 0x0: en 0x6001, ch 0x3
[2019/01/11 03:07:22:0365] USER: callback_edgeware_client: Sending PING 10...
[2019/01/11 03:07:22:0366] INFO: lws_issue_raw: ssl_capable_write (16) says 16
[2019/01/11 03:07:22:0366] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 5
[2019/01/11 03:07:22:0366] INFO: rxflow: wsi 0x2380150 change_to 1
[2019/01/11 03:07:22:0366] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 5 -> 5
[2019/01/11 03:07:22:0367] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 5 -> 1
[2019/01/11 03:07:22:0367] DEBUG: lws_handle_POLLOUT_event: 0x2380150: non mux: wsistate 0x10000117, ops ws
[2019/01/11 03:07:22:0367] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_WRITEABLE
[2019/01/11 03:07:22:0367] USER: callback_edgeware_client:  (nothing in ring)
[2019/01/11 03:07:22:0367] INFO: lws_rx_flow_control: 0x2380150 0x1
[2019/01/11 03:07:22:0367] INFO: lws_rx_flow_control: 0x2380150: bitmap 0x0: en 0x6001, ch 0x3
[2019/01/11 03:07:22:0367] INFO: rxflow: wsi 0x2380150 change_to 1
[2019/01/11 03:07:22:0367] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 1
[2019/01/11 03:07:22:0498] DEBUG: lws_read_h1: h1 path: wsi state 0x117
[2019/01/11 03:07:22:0498] EXTENSION: lws_ws_client_rx_sm:    This RX frame Final 1
[2019/01/11 03:07:22:0498] INFO: client received pong
[2019/01/11 03:07:22:0498] DEBUG:
[2019/01/11 03:07:22:0498] DEBUG: 0000: 70 69 6E 67 20 62 6F 64 79 21                      ping body!
[2019/01/11 03:07:22:0498] DEBUG:
[2019/01/11 03:07:22:0498] DEBUG: post inflate ebuf len 10
[2019/01/11 03:07:22:0498] INFO: Client doing pong callback
[2019/01/11 03:07:22:0498] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_RECEIVE_PONG (hexdump follows)
[2019/01/11 03:07:22:0498] NOTICE:
[2019/01/11 03:07:22:0499] NOTICE: 0000: 70 69 6E 67 20 62 6F 64 79 21                      ping body!
[2019/01/11 03:07:22:0499] NOTICE:
[2019/01/11 03:07:27:0373] USER: callback_edgeware_client: LWS_CALLBACK_TIMER
[2019/01/11 03:07:27:0373] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 5
[2019/01/11 03:07:27:0376] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 5 -> 1
[2019/01/11 03:07:27:0376] DEBUG: lws_handle_POLLOUT_event: 0x2380150: non mux: wsistate 0x10000117, ops ws
[2019/01/11 03:07:27:0377] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_WRITEABLE
[2019/01/11 03:07:27:0377] USER: callback_edgeware_client:  (nothing in ring)
[2019/01/11 03:07:27:0377] INFO: lws_rx_flow_control: 0x2380150 0x1
[2019/01/11 03:07:27:0377] INFO: lws_rx_flow_control: 0x2380150: bitmap 0x0: en 0x6001, ch 0x3
[2019/01/11 03:07:27:0377] USER: callback_edgeware_client: Sending PING 10...
[2019/01/11 03:07:27:0378] INFO: lws_issue_raw: ssl_capable_write (16) says 16
[2019/01/11 03:07:27:0378] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 5
[2019/01/11 03:07:27:0378] INFO: rxflow: wsi 0x2380150 change_to 1
[2019/01/11 03:07:27:0378] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 5 -> 5
[2019/01/11 03:07:27:0378] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 5 -> 1
[2019/01/11 03:07:27:0378] DEBUG: lws_handle_POLLOUT_event: 0x2380150: non mux: wsistate 0x10000117, ops ws
[2019/01/11 03:07:27:0379] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_WRITEABLE
[2019/01/11 03:07:27:0379] USER: callback_edgeware_client:  (nothing in ring)
[2019/01/11 03:07:27:0379] INFO: lws_rx_flow_control: 0x2380150 0x1
[2019/01/11 03:07:27:0379] INFO: lws_rx_flow_control: 0x2380150: bitmap 0x0: en 0x6001, ch 0x3
[2019/01/11 03:07:27:0379] INFO: rxflow: wsi 0x2380150 change_to 1
[2019/01/11 03:07:27:0379] DEBUG: _lws_change_pollfd: wsi 0x2380150: fd 7 events 1 -> 1
[2019/01/11 03:07:27:0479] DEBUG: lws_read_h1: h1 path: wsi state 0x117
[2019/01/11 03:07:27:0480] EXTENSION: lws_ws_client_rx_sm:    This RX frame Final 1
[2019/01/11 03:07:27:0480] INFO: client received pong
[2019/01/11 03:07:27:0480] DEBUG:
[2019/01/11 03:07:27:0480] DEBUG: 0000: 70 69 6E 67 20 62 6F 64 79 21                      ping body!
[2019/01/11 03:07:27:0480] DEBUG:
[2019/01/11 03:07:27:0480] DEBUG: post inflate ebuf len 10
[2019/01/11 03:07:27:0480] INFO: Client doing pong callback
[2019/01/11 03:07:27:0480] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_RECEIVE_PONG (hexdump follows)
[2019/01/11 03:07:27:0481] NOTICE:
[2019/01/11 03:07:27:0481] NOTICE: 0000: 70 69 6E 67 20 62 6F 64 79 21                      ping body!
[2019/01/11 03:07:27:0481] NOTICE:
[2019/01/11 03:07:31:1080] INFO: lws_context_destroy: ctx 0x2364160
[2019/01/11 03:07:31:1081] INFO: lws_destroy_event_pipe
[2019/01/11 03:07:31:1081] DEBUG: __remove_wsi_socket_from_fds: wsi=0x2364470, sock=5, fds pos=0, end guy pos=2, endfd=0
[2019/01/11 03:07:31:1081] INFO: __lws_close_free_wsi: 0x2380150: caller: ctx destroy
[2019/01/11 03:07:31:1081] DEBUG: __lws_close_free_wsi: 0x2380150: DROP_PROTOCOL edgeware-client
[2019/01/11 03:07:31:1081] DEBUG: __lws_close_free_wsi: real just_kill_connection: 0x2380150 (sockfd 7)
[2019/01/11 03:07:31:1081] DEBUG: __remove_wsi_socket_from_fds: wsi=0x2380150, sock=7, fds pos=0, end guy pos=1, endfd=7
[2019/01/11 03:07:31:1081] DEBUG: lwsi_set_state(0x2380150, 0x1000001e)
[2019/01/11 03:07:31:1082] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_CLOSED
[2019/01/11 03:07:31:1082] DEBUG: __lws_close_free_wsi_final: wsi 0x2380150: fd 7
[2019/01/11 03:07:31:1082] INFO: lws_vhost_unbind_wsi: vh default: count_bound_wsi 0
[2019/01/11 03:07:31:1082] DEBUG: __lws_free_wsi: 0x2380150, remaining wsi 0
[2019/01/11 03:07:31:1082] INFO: lws_vhost_destroy1
[2019/01/11 03:07:31:1082] INFO: lws_context_destroy2: ctx 0x2364160
[2019/01/11 03:07:31:1082] NOTICE: callback_rules_fifo: LWS_CALLBACK_PROTOCOL_DESTROY
[2019/01/11 03:07:31:1082] INFO: __lws_vhost_destroy2: 0x237d9c0
[2019/01/11 03:07:31:1091] INFO:   __lws_vhost_destroy2: Freeing vhost 0x237d9c0
[2019/01/11 03:07:31:1091] INFO: lws_context_destroy3: ctx 0x2364160 freed
```

**Comment by pblemel** [comment]
Thanks so much for your help.  I would really like to send you a case of beer, pizza, or cash if you can accept any of those for putting up with my self-inflicted wounds.  I would never have solved this without you, since I didn't really understand what that offending code did.  My bad, but I am learning.

Have a great rest of your night.

Peter

**Comment by lws-team** [comment]
> [2019/01/11 03:07:17:0048] USER: callback_edgeware_client: LWS_CALLBACK_CLIENT_CLOSED

This is a bug, but whether it's in lws or your code you probably need to look twice at your code first... cut and paste problem?  Missing `break;` above it so it fell through?  If not, also log the wsi `%p` and the callback reason index.

**Comment by pblemel** [comment]
I'll change the log statements to print the reason index.  Then I'll go through and compare the minimal example to my code. There's certainly a chance that I changed something while trying to troubleshoot this.

In the mean time, the callbacks in question are the last case of the switch statement and all have breaks.

```
    case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
        lwsl_err("%s: CLIENT_CONNECTION_ERROR: %s\n", __func__, in ? (char *)in : "(null)");
        vhd->client_wsi = NULL;

        // Try again.
        schedule_callback(wsi, LWS_CALLBACK_USER, 1);
        break;

    case LWS_CALLBACK_CLIENT_CLOSED:
        lwsl_user("%s: LWS_CALLBACK_CLIENT_CLOSED\n", __func__);
        if (pss->ring != NULL)
            lws_ring_destroy(pss->ring);

        vhd->client_wsi = NULL;

// Can not schedule USER callback here since at the moment the CLOSED callback is happening when we're not
// connected, resulting in a flurry of connection attempts.
//      schedule_callback(wsi, LWS_CALLBACK_USER, 1);
        break;

    case LWS_CALLBACK_USER:
        lwsl_notice("%s: LWS_CALLBACK_USER\n", __func__);

        // The USER callback is only requested when we need to (re) establish the connection.
        connect_edgeware_client(vhd);
        break;

    default:
        break;
    }

    return 0;
```

Also, now that I have the most recent LWS code building I need to update my .so to be sure this isn't something you've already fixed.

Thanks again

**Comment by pblemel** [comment]
The LWS build on the machine that I was debugging on last night (Fedora) was I'd guess about two months old. This morning I tried testing on my work desktop machine (Ubuntu) where I was building the most recent LWS repo, and the problem doesn't manifest itself there.  At this point everything works as expected. 

I'll update the Fedora machine and verify that it's not platform specific, but for now it appears to be safe to close this issue.

Thanks again,
Peter
